numpy/lib/arraysetops.py

```diff
@@ -544,6 +545,19 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
         False where an element of `ar1` is in `ar2` and True otherwise).
         Default is False. ``np.in1d(a, b, invert=True)`` is equivalent
         to (but is faster than) ``np.invert(in1d(a, b))``.
+    method : {'auto', 'sort', 'dictionary'}, optional
+        The algorithm to use. This will not affect the final result,
+        but will affect the speed. Default is 'auto'.
```

Comment:
These remarks make it sound a bit as if the only difference were speed. If I understand correctly, with the new method we're allocating an auxiliary array as large as the range of the second array, although with a 1-byte dtype. I'm not familiar with the memory need of the old method, but can it happen that the new method uses considerably more memory than the old one, which can affect users?

If I understand correctly the default `'auto'` should choose the old method for large ranges, so this probably only matters if a user goes out of their way to ask for the new method with `method='dictionary'`.

Reply:
The criterion for choosing the new method is:
$$ \log_{10}(\text{size}) > \frac{\log_{10}(\text{range}) - 2.27}{0.927}$$
which is roughly equivalent to:
$$ \text{size} > \text{range}^{1.079} / 281 $$
where `size` and `range` are of ar2. The size of the newly created matrix will have size equal to `range`. _(That_ limit is purely based on performance rather than memory.)

Therefore, in the worst case, the newly created counting array could have up to 300x the number of elements of the input array, with a 1-byte dtype.

Do you think this limit should be smaller by default? Any idea what the memory usage is for the old method?

Reply:
> Therefore, in the worst case, the newly created counting array could have up to 300x the number of elements of the input array, with a 1-byte dtype.

I think for embedded applications it could make a huge difference.

Reply:
To be clear I'm mostly asking about the docs: I think we should make it clear that the new method uses an auxiliary uint8 array as large as `arr2.ptp()`.

The criterion you posted here says that for "huge" range we need "even huger" size for `ar2`. I'd be more concerned about a "small" `ar2` array with "huge" values inside. After all, as @hameerabbasi noted, the worst-case size is approximately `np.iinfo(np.uint64).max == 18446744073709551615`. That's a lot, and independent from the size of the input arrays.

Considering this it might be wise to set a hard range cap on the `'auto'` method as well, so as not to trip up old code. But I'd leave this to the NumPy devs, I'm just the peanut gallery :)

Reply:
Let me know what you think would be a good constraint based on memory usage and I'm very happy to implement it. The inequality I posted above is purely based on performance considerations, not memory.

Also happy to describe this in the docs as well, just let me know what is decided on.

Reply:
> The criterion you posted here says that for "huge" range we need "even huger" size for ar2. I'd be more concerned about a "small" ar2 array with "huge" values inside. 

Not quite: it says that for ar2 having range $r$, you need ar2's size to be at least $0.0035$ times as big for the algorithm to be used (roughly). Then, the intermediate array will have size equal to $r$.

Reply:
A safer option might be to only create an intermediate array if it has the same size as the input arrays (rather than currently: 300x the input array size). Then at worst you are doubling memory which seems fine. Thoughts?

In the docstring I can specify `"dictionary"` method is usually much faster if the range is not too much bigger than input array sizes, but the memory usage will be equal to your data range.

Reply:
The old method, by eye, seems to have 5x to 10x the memory usage of the input arrays via intermediate allocations. Those limits could be used for complete consistency with the old method if you would like?

Reply:
> The old method, by eye, seems to have 5x to 10x the memory usage of the input arrays via intermediate allocations. 

But not using all that memory at once, right? I would assume that a single 1 GB allocation is a lot worse than ten 100 MB allocations, because the latter can reuse memory that's just been freed.

Reply:
Those variables look to be in-scope so I don't think they are freed until the function is complete, unless you re-assigned their variable names. Right? (I am not too familiar with how Python's garbage collection works)

Reply:
Yeah, if all those intermediate arrays have references (by being bound to individual names) then they stick around until the end of the function. I thought most of these were intermediate arrays in a chain of calculations that aren't even bound to a nem  (I haven't looked at the old implementation yet, sorry.)

Reply:
I will resolve this and summarize the discussion in the main thread, and ask for input.